cpp/include/raft/matrix/detail/select_k.cuh

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "select_radix.cuh"
+#include "select_warpsort.cuh"
+
+#include <raft/core/nvtx.hpp>
+#include <raft/util/mem_resource_handle.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+
+namespace raft::matrix::detail {
+
+/**
+ * Select k smallest or largest key/values from each row in the input data.
+ *
+ * If you think of the input data `in_val` as a row-major matrix with `len` columns and
+ * `batch_size` rows, then this function selects `k` smallest/largest values in each row and fills
+ * in the row-major matrix `out_val` of size (batch_size, k).
+ *
+ * @tparam T
+ *   the type of the keys (what is being compared).
+ * @tparam IdxT
+ *   the index type (what is being selected together with the keys).
+ *
+ * @param[in] in_val
+ *   contiguous device array of inputs of size (len * batch_size);
+ *   these are compared and selected.
+ * @param[in] in_idx
+ *   contiguous device array of inputs of size (len * batch_size);
+ *   typically, these are indices of the corresponding in_val.
+ * @param batch_size
+ *   number of input rows, i.e. the batch size.
+ * @param len
+ *   length of a single input array (row); also sometimes referred as n_cols.
+ *   Invariant: len >= k.
+ * @param k
+ *   the number of outputs to select in each input row.
+ * @param[out] out_val
+ *   contiguous device array of outputs of size (k * batch_size);
+ *   the k smallest/largest values from each row of the `in_val`.
+ * @param[out] out_idx
+ *   contiguous device array of outputs of size (k * batch_size);
+ *   the payload selected together with `out_val`.
+ * @param select_min
+ *   whether to select k smallest (true) or largest (false) keys.
+ * @param stream
+ * @param mr an optional memory resource to use across the calls (you can provide a large enough
+ *           memory pool here to avoid memory allocations within the call).
+ */
+template <typename T, typename IdxT>
+void select_k(const T* in_val,
+              const IdxT* in_idx,
+              size_t batch_size,
+              size_t len,
+              int k,
+              T* out_val,
+              IdxT* out_idx,
+              bool select_min,
+              rmm::cuda_stream_view stream,
+              std::optional<device_mem_resource> mr = std::nullopt)
```

Comment:
After #1111 is merged, I'm going to submit a small PR to add this device memory resource to the new "resource container" API so we won't need to explicitly accept this argument in these functions.

Reply:
Adding to my comment above- the `rmm::device_memory_resource` won't need to be accepted in the public API functions. The functions in the `detail` API can accept the `rmm::device_memory_resource` as usual. We also won't need specialized factory functions to construct them. To start, we can add an option in the `raft::device_resources` (which the `raft::handle_t` inherits for backwards compatibility) which populates the `rmm::device_memory_resource` internally. The goal is to eventually get to the point where all of the primitives are just accepting `raft::resources` instances and they are then using the corresponding accessor functions to get the resources they want- which alleviates the need to include everything everytime in the primitives.

Reply:
Hmm, I'm a bit puzzled now. While I'm very looking forward toward using the new `raft::resources`, I'm not sure if it's applicable here.
As far as I understand, the `raft::resources` is supposed to give access to singleton-like resources, the ones that are shared among all use sites, right? Then, I don't see a difference between using the memory resource taken from `raft::resources` and the one that is set as the default using the rmm api. Both of them have the semantics of one-default-resource. However, what we provide here in the arguments, is the option to decide at the call site, whether we want to use the default shared resource, or a small dedicated resource (e.g. a pre-allocated memory pool, to avoid any cuda allocation calls within the scope of the function; or a managed memory to save some gpu ram). How can we provide this functionality with `raft::resources`? Perhaps, we could have a few different, named memory resources, or a memory-resource-pool? Still, this doesn't give the control over the decision at the call site.

Reply:
As far as I'm concerned, we are talking about a secondary memory resource that is used explicitly for temporary workspace allocations so I'm not sure why it matters whether this value is set at the call site or in the handle/resources before invoking the function. There are several other places in RAFT which can also make use of such a pool or managed (or managed pool) resource for temporary workspace allocations and whether or not users want to 1) set a new one on the handle before each invocation of a function that uses it, 2) whether they just want to set one and let all the calls use that resource, or 3) whether they just want to reuse the default resource, doesn't seem particularly relevant to this discussion. The primary goal here is to decouple and remove resource configuration from the algorithm contract and allow the resource configuration information to propagate from the handle/resources objects.

Reply:
The main use-case I have in mind (besides preallocating a pool before calling ivf_pq::search and the likes), is when, while developing a new algorithm, I know that a particular helper function may require a big temporary buffer for certain inputs. In this case, I may want to make this helper use the managed memory, while keeping other functions using the device memory only (for performance reasons). In other use case I could be confident that the required workspace size is small enough that I don't need to resort to managed memory.
 I guess, that sounds like your option (1)? I'm concerned that this state-machine approach of setting the new resource before and after function invocation may be more error prone and I'm not sure how it would behave in the presence of cuda streams without explicit synchronization (can the unset memory resource be destroyed before the kernels inside the function are finished?..).

Reply:
From what you've described so far, I'm still not seeing the need to even use an allocator in addition to a single "temporary workspace allocator". If the algorithm functions include the temporary workspace requirements in their documentation, the user should be able to make the decision as to whether they can get away with a small pooled resource, whether they could do a pooled resource that uses managed memory, or whether they are better doing a managed resource without a pool. Also keep in mind, this is in addition to the standard default allocator, of which the user can also choose between a pooled, managed, or managed pool allocator (which I suspect they are more likely to change anyways in applications where they need to oversubscribe the GPU memory).

This still makes me wonder upon which circumstances they would need more than 2 separate resources- one for standard memory allocations (and temporary allocations which don't use a second allocator) and one which is the "preferred" source for temporary allocations which might benefit from a secondary resource.

The general rule that users tend to go by for the default allocator is "if you don't have enough GPU memory and need to oversubscribe (pending caveats), you can switch your default allocator to a managed one." The same applies, "you can get a performance boost with pooled allocator if you have the memory for it", and finally, "managed pooled allocator can provide the best of both worlds if you don't need to oversubscribe by too much".

Reply:
Another thing to consider here is that multiple handles/device_resources/resources instances can be used. For any extreme cases where a user feels they absolutely must use a pool allocator for the temp workspace allocator in one function invocation and a managed allocator in another, they can simply create 2 handles up front that both use the same user stream (and potentially even the same stream pool) but different temp workspace allocators. This would still afford them the flexibility of configuring the resources to their liking while still decoupling the resource configuration from the algorithm's contract.

Reply:
Ok. Then, maybe shall I just remove the memory resource argument from the public API, and then we could migrate all use cases in detail namespaces in a separate PR? Otherwise, I'd have to make here a lot of logic code changes in ifv-* algorithms.

Reply:
Yep, I don't think we need to change them all right here but I wanted to catch this one before it goes "live" and while we finally have the new resources API ready.

Also, with regards to your note about the resources being singletons, I agree the verbiage of the docs in the resources API is misleading there- these resources are not global or static singletons, they are really just members of an instance of a resources object. Some of the verbiage there is left over from the original docs on the handle_t but I'll adjust them to make this more clear.

Reply:
#1111 was just merged so I'll open up a PR tomorrow to add the `workspace_resource` to the resources API, with corresponding mutators and accessors. Likely, I'll add an argument to the `device_resources` constructor to set the optional resource for now (as a shared_ptr).

Reply:
I've opened https://github.com/rapidsai/raft/pull/1137 with the change so users will be able to create a `raft::device_resources`/`handle_t` w/ an optional `rmm::mr::device_memory_resource` and then we can utilize that memory resource w/ the passed in `device_resources` object (which is now a superclass of `raft::handle_t` with the following:

```c++
#include <raft/core/resource/device_memory_resource.hpp>
...
rmm::mr::device_memory_resource *mr = get_workspace_resource(handle);
```

You can also access it directly w/ the `device_resources` instance: `handle.get_workspace_resource()`, but I'm hoping to see us eventually stop using that object directly and start accepting `raft::resources` at some point in the future so that primitives don't have to include headers for resources that aren't being used.